File: src/leetcode_train/PrintTree.java
package leetcode_train;
import java.util.*;
import java.math.*;

/**
 * Created by EVERCX on 2018/3/28.
 *
 * https://leetcode-cn.com/problems/print-binary-tree/description/
 */



class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}


public class PrintTree {

    public List<List<String>> printTree(TreeNode root) {

        int maxDeep = maxDeep(root);
        int n  = (int)Math.pow(2,maxDeep) - 1;  //二维数组的列数

        List<List<String>> result = new ArrayList<List<String>>();

        for(int i=0;i < maxDeep;i++){
            List<String> curList = new ArrayList<String>();
            for(int j = 0;j < n; j++){
                curList.add("");
            }

            result.add(curList);
        }

        setNodesToList(result,root,0,0,n);
        return result;
    }

    public void setNodesToList(List<List<String>> result,TreeNode node,int listLevel,int leftIndex,int rightIndex){

        if(node == null) return;

        int index = (leftIndex + rightIndex) / 2;

        result.get(listLevel).set(index,node.val+"");

        setNodesToList(result,node.left,listLevel+1,leftIndex,index-1);

        setNodesToList(result,node.right,listLevel+1,index+1,rightIndex);

    }

    public int maxDeep(TreeNode root){
        if(root == null) return 0;
        int leftDeep =  maxDeep(root.left);
        int rightDeep = maxDeep(root.right);

        return (leftDeep > rightDeep ? leftDeep:rightDeep) + 1;
    }

    public static void main(String[] args) {

        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.left.right = new TreeNode(4);
        root.right = new TreeNode(3);

        List<List<String>> result = new PrintTree().printTree(root);

        List<String> r = new ArrayList<>();

        r.add("1");
        r.add("");
        r.add(0,"");

        System.out.println(result.toString());
        System.out.println(new PrintTree().printTree(root));

    }
}

File: leetcode_train_by_py/printTree.py


# https://leetcode.com/problems/print-binary-tree/description/

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def printTree(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[str]]
        """

        deep = self.deepTree(root)
        print("deep",deep)

        # 结果数组的尺寸

        # 每行的元素个数
        number = pow(2,deep)-1

        result = [ [ "" for j in range(number)] for i in range(deep)]

        self.dfs(root,0,result,0,number-1)

        return result



    def dfs(self,node,deep,result,left,right):

        mid = int((left + right) / 2)
        result[deep][mid] = str(node.val)

        if node.left:
            self.dfs(node.left,deep+1,result,left,mid-1)
        if node.right:
            self.dfs(node.right,deep+1,result,mid+1,right)

    def deepTree(self,root):

        if root == None:
            return 0

        left_deep = self.deepTree(root.left) + 1
        right_deep = self.deepTree(root.right) + 1

        return max(left_deep,right_deep)




root = TreeNode("1")
root.left = TreeNode("2")
root.left.right = TreeNode("4")
root.right = TreeNode("3")


result = Solution().printTree(root)

for i in range(len(result)):
    print(result[i])

